### PR TITLE
Pause and resume rule group evaluation

### PR DESCRIPTION
**Pull request by rajagopalanand**
This PR is for https://github.com/prometheus/prometheus/issues/13630

### CODE REVIEW

File: rules/group.go
Comment (emanlodovice) [inline rules/group.go:234]: Should this be moved inside the `if missed > 0` block? No point in calling `Add` if `missed` is 0.

File: rules/group.go
Comment (emanlodovice) [inline rules/group.go:328]: Can we keep the comment
```
		// have updated the latest values, on which some alerts might depend.
                ...
```
This is useful information.

File: rules/group.go
Comment (emanlodovice) [inline rules/group.go:101]: Nit: I think this needs a better name because `EvaluateRuleGroup` sounds like a function that performs that evaluation, not a functions that decides whether to evaluate or not. 
Suggestion: `RuleGroupEvalControlFunc`

File: rules/manager_test.go
Comment (emanlodovice) [inline rules/manager_test.go:320]: Can we assert that the rule did not evaluate when we switched it to false?

File: rules/group.go
Comment (rajagopalanand) [reply rules/group.go]: Yes, done

File: rules/group.go
Comment (rajagopalanand) [reply rules/group.go:332]: Good point! Done

File: rules/manager_test.go
Comment (rajagopalanand) [reply rules/manager_test.go:321]: Done

File: rules/group.go
Comment (bwplotka) [inline rules/group.go:45]: Do you mind adding comment on this exported struct?

File: rules/group.go
Comment (bwplotka) [inline rules/group.go:113]: I know other fields are not commented, but can we at least break this pattern and explain how this options should be used by the callers? What is the contract here? What could be surprising?

File: rules/group.go
Comment (bwplotka) [inline rules/group.go:218]: One consequence of removing this one, is that we restore only after SECOND eval after time ticks. I think that's fine,given commentary... We do exactly that 🤔 Curious why we didn't do that in the first place!

File: rules/manager.go
Comment (bwplotka) [inline rules/manager.go:90]: Can we add comment?  Also name is wrong, this should be perhaps:
```suggestion
func AlwaysTrueShouldEvaluateFunc(ctx context.Context, g *Group, evalTimestamp time.Time, doneCh <-chan struct{}) bool {
```
WDYT?

**Review by bwplotka — DISMISSED** [review]
Thanks! I know there are some opinions to not do it (complexity that Prometheus will not use), but I would love more help (potential for maintainera) here and more users/activity so reviewed.

Generally LGTM, just some suggestions, mostly naming. Thanks! Can we undraft?

cc @juliusv @beorn7

File: rules/manager.go
Comment (rajagopalanand) [reply rules/manager.go:90]: Like this! Will make the changes soon

File: rules/group.go
Comment (rajagopalanand) [reply rules/group.go:45]: I'm going to make this struct private. After looking at this again, I don't see why this has to be exported

**Review by emanlodovice — APPROVED** [review]
lgtm

File: rules/group.go
Comment (rapphil) [inline rules/group.go:104]: I think in [this comment](https://github.com/prometheus/prometheus/pull/14142/files#r1624224129) the suggestion was to change the name of the default function, and not the type.

`DefaultEvaluateRuleGroupFunc` -> `AlwaysTrueShouldEvaluateFunc`

File: rules/manager_test.go
Comment (rapphil) [inline rules/manager_test.go:309]: Since you have control over the `evalIterationFunc`, you can use WaitGroups instead of using sleeps, which can make tests flaky and longer than we need.

File: rules/group.go
Comment (juliusv) [inline rules/group.go:251]: I see that some of these methods are exported (capitalized), others are not. Are the exported ones used from the outside by your code? Otherwise I'd keep them all unexported to make it clear that that's not the case.

And just to check that I am understanding things correctly, are these mutex-protected accessor functions only needed for testing, not because of concurrency protection within the rule manager itself?

**Review by juliusv — COMMENTED** [review]
My general opinion from https://github.com/prometheus/prometheus/issues/13630#issuecomment-2011666548 about taking on major code that we don't use in Prometheus still stands, but in case others overrule me, just a few comments on the code itself.

**Comment by beorn7** [comment]
Not a code level review comment, just about the general concern of dead code: I share @juliusv's concerns. But I see a bunch of "soft" reasons to accept this PR:
- @bwplotka pretty clearly said "yes", while nobody expressed a strong "no". And we even discussed this particular case at the dev summit, so we know that a bunch of people have looked at this. Generally, I think we should lean towards a "yes" in those cases. (Note that this is not a irrevocable decision. It's not a user-facing feature. If we realize later that we are drowning in dead code that nobody feels responsible for, we can still rip it out again without further ado.)
- It's not unusual that we get asked to add some code that is dead code for vanilla Prometheus to accommodate needs of sibling projects like Cortex/Thanos/Mimir. I certainly believe that we should not reject _all_ of those requests (as we did long ago), but that we should use "best judgement" what's still OK. I think we have to fine-tune the exact threshold, and as long as we haven't reached the limit, we kind of _have_ to be on the more lenient side, until we do notice the limit. This also implies that we might push back harder in the future. Being lenient also helps avoiding "preferential treatment" of one sibling project over another, based on the bias that everyone of us will have (e.g. I am in touch with Mimir devs in my day job, so their problems might appear more relevant to me than those of Cortex or Thanos devs). But as said, all of that "within reason".
- Finally, Cortex folks are not just dropping this here and then they'll disappear. On the contrary, the intention was stated that they plan to engage even more. And that circles back to what @bwplotka said. In other words: Yes, this PR creates a maintenance burden, but if it also leads to Cortex folks providing more maintenance work in upstream Prometheus, it's a net plus.

File: rules/group.go
Comment (rajagopalanand) [reply rules/group.go:251]: I am changing this to `requiresRestoration()`. I cannot keep the field and method name the same 

> And just to check that I am understanding things correctly, are these mutex-protected accessor functions only needed for testing, not because of concurrency protection within the rule manager itself?

Correct, needed only for testing

File: rules/manager_test.go
Comment (rajagopalanand) [reply rules/manager_test.go]: I changed the test to use channels instead of `time.Sleep`

**Comment by juliusv** [comment]
Ok, I don't want to block it if there are other team members who feel like it's worth maintaining it in Prometheus. My concerns stand, but it's always a tradeoff / "best judgement" kind of thing without a 100% clear answer. If it does attract more Cortex developer participation that benefits Prometheus itself in the future, then all the better. :+1: otherwise overall from a code standpoint.

**Comment by rajagopalanand** [comment]
Can this be merged?

**Comment by beorn7** [comment]
@bwplotka you looked at the code before. Could you make the final call here and re-approve/merge?

**Review by bwplotka — COMMENTED** [review]
Hi, sorry for lack of response and thanks for patience and work here. Was busy, but also honestly speaking, bit more sceptical about this change given:

A. valid comments from @juliusv
B. increasing complexity of the approach in this PR after closer look (see questions in Unblocking You section).
C. Lack of good process of Prometheus ensuring we don't kill/break the code Prometheus does not use by accident later on. Ideally we don't have later on angry Cortex users/maintainers when we accidently do this.
D. I'm not sure I agree with the HA rule eval design that triggers this change. You don't need to agree with me and I don't know all requirements, but something feels off. It feels easier to dedup on write instead of trying, with strong consistency turn off/turn on (passive active) ruler replicas. So not only this use case is remote to Prometheus, but is also something I would not recommend doing.. thus makes me bit hesitant to put green mark on, especially with big complexity.

> Finally, Cortex folks are not just dropping this here and then they'll disappear. On the contrary, the intention was stated that they plan to engage even more. And that circles back to what @bwplotka said. In other words: Yes, this PR creates a maintenance burden, but if it also leads to Cortex folks providing more maintenance work in upstream Prometheus, it's a net plus.

+1, although one question is.. what blocks e.g. you @rajagopalanand from helping on some Prometheus duties (reviews/ issues) without this merged for now? Like it's normal sometimes to maintain light fork for Thanos/Cortex/AWS/Google/Grafana specific changes anyway and sometimes it's an ok choice too.

### Unblocking You

To sum up, I assuming positive intent, but I would love to challenge this PR again, because it feels there might be "lighter" alternative to fulfill Cortex HA rule eval use case, thus allowing extensibility for little change on Prometheus side. Feel free to disagree! 

In my eyes the lightest way to unblock you is:

1. Ensure callers  can unload/load certain groups on demand. e.g.  `Load/Create` or`Unload/Stop` methods. In fact we have Load and stop already, so it feels like a small change.
2. Ensure callers can have strong consistency of bulk unloading vs evaluation. This is what the `ShouldEvaluateRuleGroupFunc` was intended to do, but do we even need this in Prometheus code? Can we just make [`evalIterationFunc`](https://github.com/prometheus/prometheus/blob/cb7306155bf7409d5f5fcc41cc71ecb81e380666/rules/group.go#L74) public/wrappable?

Some specific actions to address perhaps:

* @juliusv is [right](https://github.com/prometheus/prometheus/pull/14142#discussion_r1627058093). Big reset method with lot's o boilerplate is prone to mistakes in future. Reset also feels like a duplicate, because we already kind of "reset" when doing "stop" on  load/unload method https://github.com/prometheus/prometheus/blob/cb7306155bf7409d5f5fcc41cc71ecb81e380666/rules/manager.go#L193, no? What blocks us from having one "stop", one way of recreating the recording rule? If we need to "pause/cache" some state, can we do it on caller side (Cortex). Ideally we just (1) above.
* `ShouldEvaluateRuleGroupFunc` is a bit of code smell at this point. It's a generic (another) function that passes the whole context, not only to check if things should be evaluated but also to deactivate/activate. If nothing else, name ideally changes. But better, can we try doing (2), so just hook to something existing and rule/manager will just evaluate and code for blocking/ unloading/loading is on Cortex side? 
* Can we address comments? I see some questions/comments I made or others made and they are still dangling. Do you mind addressing them (agree/disagree/done(resolved))?
* Can we add clear `NOTE(@rajagopalanand): This method is used for Cortex HA rule case. Consult/cc Cortex team when changing.` comment on public methods Cortex will use, but Prometheus does not?

Sorry for overloading this, we might have a different opinions/coding style in Prometheus too perhaps 🤗 I am sure we can find some consensus in this case.

**Comment by rajagopalanand** [comment]
Thank you for the reviews and time. I understand the issues with this PR and therefore going to withdraw/close the PR

**Pull request closed without merging**
